File: app/src/test/java/com/productivity/habits/domain/gamification/GamificationEngineTest.kt
package com.productivity.habits.domain.gamification

import com.google.common.truth.Truth.assertThat
import com.productivity.habits.data.local.entity.HabitEntity
import com.productivity.habits.data.local.entity.HabitFrequencyType
import com.productivity.habits.data.local.entity.HabitLogEntity
import com.productivity.habits.data.local.entity.HabitTargetType
import org.junit.Test
import java.time.Instant
import java.util.UUID

class GamificationEngineTest {

    @Test
    fun calculateStreakMultiplier_returnsCorrectMultiplierForStreakLengths() {
        assertThat(GamificationEngine.calculateStreakMultiplier(0)).isEqualTo(1.0)
        assertThat(GamificationEngine.calculateStreakMultiplier(6)).isEqualTo(1.0)
        assertThat(GamificationEngine.calculateStreakMultiplier(7)).isEqualTo(1.25)
        assertThat(GamificationEngine.calculateStreakMultiplier(13)).isEqualTo(1.25)
        assertThat(GamificationEngine.calculateStreakMultiplier(14)).isEqualTo(1.5)
        assertThat(GamificationEngine.calculateStreakMultiplier(29)).isEqualTo(1.5)
        assertThat(GamificationEngine.calculateStreakMultiplier(30)).isEqualTo(2.0)
        assertThat(GamificationEngine.calculateStreakMultiplier(100)).isEqualTo(2.0)
    }

    @Test
    fun xpThresholdForLevel_quadraticCurveProgression() {
        assertThat(GamificationEngine.xpThresholdForLevel(1)).isEqualTo(0L)
        assertThat(GamificationEngine.xpThresholdForLevel(2)).isEqualTo(100L)
        assertThat(GamificationEngine.xpThresholdForLevel(3)).isEqualTo(250L)
        assertThat(GamificationEngine.xpThresholdForLevel(4)).isEqualTo(450L)
        assertThat(GamificationEngine.xpThresholdForLevel(5)).isEqualTo(700L)
        assertThat(GamificationEngine.xpThresholdForLevel(10)).isEqualTo(2700L)
        assertThat(GamificationEngine.xpThresholdForLevel(20)).isEqualTo(10450L)
    }

    @Test
    fun calculateProgression_resolvesLevelTitlesAndProgressFractions() {
        // Level 1: 0 XP
        val prog1 = GamificationEngine.calculateProgression(totalXp = 0L)
        assertThat(prog1.level).isEqualTo(1)
        assertThat(prog1.title).isEqualTo(PlayerTitle.NOVICE)
        assertThat(prog1.currentLevelBaseXp).isEqualTo(0L)
        assertThat(prog1.nextLevelTargetXp).isEqualTo(100L)
        assertThat(prog1.progressFraction).isEqualTo(0.0f)

        // Level 1: 50 XP (halfway to Lv 2)
        val progHalf = GamificationEngine.calculateProgression(totalXp = 50L)
        assertThat(progHalf.level).isEqualTo(1)
        assertThat(progHalf.progressFraction).isEqualTo(0.5f)

        // Level 5: 700 XP (Apprentice)
        val prog5 = GamificationEngine.calculateProgression(totalXp = 700L, longestActiveStreak = 14)
        assertThat(prog5.level).isEqualTo(5)
        assertThat(prog5.title).isEqualTo(PlayerTitle.APPRENTICE)
        assertThat(prog5.activeStreakMultiplier).isEqualTo(1.5)

        // Level 10: 2700 XP (Pathfinder)
        val prog10 = GamificationEngine.calculateProgression(totalXp = 2700L, longestActiveStreak = 30)
        assertThat(prog10.level).isEqualTo(10)
        assertThat(prog10.title).isEqualTo(PlayerTitle.PATHFINDER)
        assertThat(prog10.activeStreakMultiplier).isEqualTo(2.0)

        // Level 20: 10450 XP (Grandmaster)
        val prog20 = GamificationEngine.calculateProgression(totalXp = 10450L)
        assertThat(prog20.level).isEqualTo(20)
        assertThat(prog20.title).isEqualTo(PlayerTitle.GRANDMASTER)
    }

    @Test
    fun calculateHabitDayBaseXp_booleanHabit() {
        val now = Instant.now()
        val habit = HabitEntity(
            id = "h1",
            title = "Drink Water",
            color = "#000000",
            frequencyType = HabitFrequencyType.DAILY,
            targetType = HabitTargetType.BOOLEAN,
            createdAt = now,
            updatedAt = now
        )
        val log = HabitLogEntity(
            id = UUID.randomUUID().toString(),
            habitId = "h1",
            date = "2026-08-17",
            timestamp = now,
            completed = true,
            createdAt = now,
            updatedAt = now
        )

        val xpCompleted = GamificationEngine.calculateHabitDayBaseXp(habit, listOf(log), isCompleted = true)
        assertThat(xpCompleted).isEqualTo(GamificationEngine.BASE_BOOLEAN_XP)

        val xpNotCompleted = GamificationEngine.calculateHabitDayBaseXp(habit, emptyList(), isCompleted = false)
        assertThat(xpNotCompleted).isEqualTo(0)
    }

    @Test
    fun calculateHabitDayBaseXp_timerHabit() {
        val now = Instant.now()
        val habit = HabitEntity(
            id = "h2",
            title = "Deep Work",
            color = "#000000",
            frequencyType = HabitFrequencyType.DAILY,
            targetType = HabitTargetType.TIMER,
            targetValue = 25.0,
            createdAt = now,
            updatedAt = now
        )
        val log = HabitLogEntity(
            id = UUID.randomUUID().toString(),
            habitId = "h2",
            date = "2026-08-17",
            timestamp = now,
            completed = true,
            durationSeconds = 1500L, // 25 minutes
            createdAt = now,
            updatedAt = now
        )

        val xp = GamificationEngine.calculateHabitDayBaseXp(habit, listOf(log), isCompleted = true)
        // 25 mins (25 XP) + 10 XP bonus
        assertThat(xp).isEqualTo(35)
    }

    @Test
    fun calculateHabitDayBaseXp_slotsHabit() {
        val now = Instant.now()
        val habit = HabitEntity(
            id = "h3",
            title = "Walks",
            color = "#000000",
            frequencyType = HabitFrequencyType.TIMES_PER_DAY,
            timesPerDay = 3,
            targetType = HabitTargetType.BOOLEAN,
            createdAt = now,
            updatedAt = now
        )
        val logs = (0 until 3).map { slot ->
            HabitLogEntity(
                id = UUID.randomUUID().toString(),
                habitId = "h3",
                date = "2026-08-17",
                timestamp = now,
                intervalIndex = slot,
                completed = true,
                createdAt = now,
                updatedAt = now
            )
        }

        val xp = GamificationEngine.calculateHabitDayBaseXp(habit, logs, isCompleted = true)
        // 3 slots * 10 XP + 15 XP bonus = 45 XP
        assertThat(xp).isEqualTo(45)
    }

    @Test
    fun applyMultiplier_scalesXpCorrectly() {
        assertThat(GamificationEngine.applyMultiplier(20, 1.0)).isEqualTo(20L)
        assertThat(GamificationEngine.applyMultiplier(20, 1.25)).isEqualTo(25L)
        assertThat(GamificationEngine.applyMultiplier(20, 1.5)).isEqualTo(30L)
        assertThat(GamificationEngine.applyMultiplier(20, 2.0)).isEqualTo(40L)
    }
}

File: test/domain/gamification_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:habit_tracker/domain/gamification/gamification_engine.dart';
import 'package:habit_tracker/domain/gamification/player_title.dart';
import 'package:habit_tracker/domain/models/habit.dart';
import 'package:habit_tracker/domain/models/habit_frequency_type.dart';
import 'package:habit_tracker/domain/models/habit_log.dart';
import 'package:habit_tracker/domain/models/habit_target_type.dart';
import 'package:uuid/uuid.dart';

void main() {
  const uuid = Uuid();

  test('calculateStreakMultiplier_returnsCorrectMultiplierForStreakLengths', () {
    expect(GamificationEngine.calculateStreakMultiplier(0), 1.0);
    expect(GamificationEngine.calculateStreakMultiplier(6), 1.0);
    expect(GamificationEngine.calculateStreakMultiplier(7), 1.25);
    expect(GamificationEngine.calculateStreakMultiplier(13), 1.25);
    expect(GamificationEngine.calculateStreakMultiplier(14), 1.5);
    expect(GamificationEngine.calculateStreakMultiplier(29), 1.5);
    expect(GamificationEngine.calculateStreakMultiplier(30), 2.0);
    expect(GamificationEngine.calculateStreakMultiplier(100), 2.0);
  });

  test('xpThresholdForLevel_quadraticCurveProgression', () {
    expect(GamificationEngine.xpThresholdForLevel(1), 0);
    expect(GamificationEngine.xpThresholdForLevel(2), 100);
    expect(GamificationEngine.xpThresholdForLevel(3), 250);
    expect(GamificationEngine.xpThresholdForLevel(4), 450);
    expect(GamificationEngine.xpThresholdForLevel(5), 700);
    expect(GamificationEngine.xpThresholdForLevel(10), 2700);
    expect(GamificationEngine.xpThresholdForLevel(20), 10450);
  });

  test('calculateProgression_resolvesLevelTitlesAndProgressFractions', () {
    // Level 1: 0 XP
    final prog1 = GamificationEngine.calculateProgression(totalXp: 0);
    expect(prog1.level, 1);
    expect(prog1.title, PlayerTitle.novice);
    expect(prog1.currentLevelBaseXp, 0);
    expect(prog1.nextLevelTargetXp, 100);
    expect(prog1.progressFraction, 0.0);

    // Level 1: 50 XP (halfway to Lv 2)
    final progHalf = GamificationEngine.calculateProgression(totalXp: 50);
    expect(progHalf.level, 1);
    expect(progHalf.progressFraction, 0.5);

    // Level 5: 700 XP (Apprentice)
    final prog5 = GamificationEngine.calculateProgression(totalXp: 700, longestActiveStreak: 14);
    expect(prog5.level, 5);
    expect(prog5.title, PlayerTitle.apprentice);
    expect(prog5.activeStreakMultiplier, 1.5);

    // Level 10: 2700 XP (Pathfinder)
    final prog10 = GamificationEngine.calculateProgression(totalXp: 2700, longestActiveStreak: 30);
    expect(prog10.level, 10);
    expect(prog10.title, PlayerTitle.pathfinder);
    expect(prog10.activeStreakMultiplier, 2.0);

    // Level 20: 10450 XP (Grandmaster)
    final prog20 = GamificationEngine.calculateProgression(totalXp: 10450);
    expect(prog20.level, 20);
    expect(prog20.title, PlayerTitle.grandmaster);
  });

  test('calculateHabitDayBaseXp_booleanHabit', () {
    final now = DateTime.now().toUtc();
    final habit = Habit(
      id: 'h1',
      title: 'Drink Water',
      color: '#000000',
      frequencyType: HabitFrequencyType.daily,
      targetType: HabitTargetType.boolean,
      createdAt: now,
      updatedAt: now,
    );
    final log = HabitLog(
      id: uuid.v4(),
      habitId: 'h1',
      date: '2026-08-17',
      timestamp: now,
      completed: true,
      createdAt: now,
      updatedAt: now,
    );

    final xpCompleted = GamificationEngine.calculateHabitDayBaseXp(habit, [log], true);
    expect(xpCompleted, GamificationEngine.baseBooleanXp);

    final xpNotCompleted = GamificationEngine.calculateHabitDayBaseXp(habit, [], false);
    expect(xpNotCompleted, 0);
  });

  test('calculateHabitDayBaseXp_timerHabit', () {
    final now = DateTime.now().toUtc();
    final habit = Habit(
      id: 'h2',
      title: 'Deep Work',
      color: '#000000',
      frequencyType: HabitFrequencyType.daily,
      targetType: HabitTargetType.timer,
      targetValue: 25.0,
      createdAt: now,
      updatedAt: now,
    );
    final log = HabitLog(
      id: uuid.v4(),
      habitId: 'h2',
      date: '2026-08-17',
      timestamp: now,
      completed: true,
      durationSeconds: 1500, // 25 minutes
      createdAt: now,
      updatedAt: now,
    );

    final xp = GamificationEngine.calculateHabitDayBaseXp(habit, [log], true);
    // 25 mins (25 XP) + 10 XP bonus
    expect(xp, 35);
  });

  test('calculateHabitDayBaseXp_slotsHabit', () {
    final now = DateTime.now().toUtc();
    final habit = Habit(
      id: 'h3',
      title: 'Walks',
      color: '#000000',
      frequencyType: HabitFrequencyType.timesPerDay,
      timesPerDay: 3,
      targetType: HabitTargetType.boolean,
      createdAt: now,
      updatedAt: now,
    );
    final logs = List.generate(3, (slot) {
      return HabitLog(
        id: uuid.v4(),
        habitId: 'h3',
        date: '2026-08-17',
        timestamp: now,
        intervalIndex: slot,
        completed: true,
        createdAt: now,
        updatedAt: now,
      );
    });

    final xp = GamificationEngine.calculateHabitDayBaseXp(habit, logs, true);
    // 3 slots * 10 XP + 15 XP bonus = 45 XP
    expect(xp, 45);
  });

  test('applyMultiplier_scalesXpCorrectly', () {
    expect(GamificationEngine.applyMultiplier(20, 1.0), 20);
    expect(GamificationEngine.applyMultiplier(20, 1.25), 25);
    expect(GamificationEngine.applyMultiplier(20, 1.5), 30);
    expect(GamificationEngine.applyMultiplier(20, 2.0), 40);
  });
}
